### DISCUSSION TITLE
PHP Symfony How To Make CMS Modular

### DISCUSSION THREAD

**Discussion by N4thyra** [symfony/symfony#58468] [Q&A] [answered]
I’m new to Symfony, I’ve been experimenting with the framework and I kind of like it.
My goal is to create a CMS that is modular (extensible).

The core should contain stuff such as:
1. User management
2. Routes
3. Pages (articles)
4. Media

Modules (extensions) could be e.g.:
1. Products
2. References
3. Distributors
4. Newsletters

I’ve been reading about hexagonal architecture lately, and it seems like a good idea to take this path. Although you can find several example projects on GitHub and other places, they are were simplified.
I haven’t seen any Symfony CMS using this architecture where I could take inspiration from.

Sylius takes the advantage of bundles, is it a good way to go?

Basically, I would want to achieve the following:
1. The core should be decoupled from modules
2. Modules should be able to communicate between one another
3. It would be nice to be able to fix bugs or add new features to the core or modules without breaking everything

Could an experienced Symfony developer give me a piece of advice on the “best” approach?

**Comment by jbcr** [comment, marked as answer]
Hi,

Like Symfony and Sylius, you can write independent components and bundle to wrap the components into Symfony.

Sylius has some bundles and the CoreBundle makes relations between all other bundles and components to make a consistent solution.

Each component can be used independently (without dependency on another component), use the interface from the contacts library, and can be replaced by another implementation if necessary.

**Reply by N4thyra** [reply]
Hey,

thank you for your advice I appreciate it.

I was quite surprised to see that the documentation of Symfony discourages people from using bundles, but on the other hand, most large CMS projects use them.

In programming, there are many ways to achieve what's needed, and I don't intend to reinvent the wheel as I don't want to get burnt.

**Reply by stof** [reply]
The documentation of Symfony discourages using bundles for the project code, because this is overkill for most projects.
Sylius uses bundles because they ship bundles as **separate** package, not as part of a single monolith package containing multiple bundles.

**Reply by N4thyra** [reply]
Understood, everything is now a lot clearer. 
Thanks @jbcr & @stof! 
You've been of great help.
